catchup/classBasedPeerSelector.go

```diff
@@ -0,0 +1,158 @@
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package catchup
+
+import (
+	"errors"
+	"github.com/algorand/go-algorand/network"
+	"github.com/algorand/go-deadlock"
+	"sort"
+	"time"
+)
+
+// The duration after which we reset the downloadFailures for a rankPooledPeerSelector
+const lastCheckedDuration = 10 * time.Minute
+
+// classBasedPeerSelector is a rankPooledPeerSelector that tracks and ranks classes of peers based on their response behavior.
+// It is used to select the most appropriate peers to download blocks from - this is most useful when catching up
+// and needing to figure out whether the blocks can be retrieved from relay nodes or require archive nodes.
+type classBasedPeerSelector struct {
+	mu            deadlock.Mutex
+	peerSelectors []*wrappedPeerSelector
+}
+
+func makeClassBasedPeerSelector(peerSelectors []*wrappedPeerSelector) *classBasedPeerSelector {
+	// Sort the peerSelectors by priority
+	sort.SliceStable(peerSelectors, func(i, j int) bool {
+		return peerSelectors[i].priority < peerSelectors[j].priority
+	})
+	return &classBasedPeerSelector{
+		peerSelectors: peerSelectors,
+	}
+}
+
+func (c *classBasedPeerSelector) rankPeer(psp *peerSelectorPeer, rank int) (int, int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	peerSelectorSortNeeded := false
+	poolIdx, peerIdx := -1, -1
+	for _, wp := range c.peerSelectors {
+		// See if the peer is in the class, ranking it appropriately if so
+		poolIdx, peerIdx = wp.peerSelector.rankPeer(psp, rank)
+		if poolIdx < 0 || peerIdx < 0 {
+			// Peer not found in this class
+			continue
+		}
+		// Peer was in this class, if there was any kind of download issue, we increment the failure count
+		if rank >= peerRankNoBlockForRound {
+			wp.downloadFailures++
+		}
+
+		// If we have failed more than the tolerance factor, we re-sort the slice of peerSelectors
+		if wp.downloadFailures > wp.toleranceFactor {
+			peerSelectorSortNeeded = true
+		}
+		break
+	}
+
+	if peerSelectorSortNeeded {
+		c.sortPeerSelectors()
+	}
+
+	return poolIdx, peerIdx
+}
+
+// sortPeerSelectors sorts the peerSelectors by tolerance factor violation, and then by priority
+// It should only be called within a locked context
+func (c *classBasedPeerSelector) sortPeerSelectors() {
+	psUnderTolerance := make([]*wrappedPeerSelector, 0, len(c.peerSelectors))
```

Comment:
how about the following instead of sorting the `peerSelectors` slice:
1. Threat position in a slice as a priority (order) to try - this is similar to the original peer selector
2. Use failures count and `lastCheckedTime` resetting logic as a "disable" flag.
3. Each getNextPeer() call simply goes over the slice and calls `getNextPeer` from the selector in a first active (not disabled) slice position.

Reply:
Updated in the latest push, think we are on same page.